Bp deep dive (#412)
leaningtech/labs: sites/labs/src/content/blog/browserpod-deep-dive.mdx

The disk and network backends of BrowserPod are especially important for compatibility with a broad range of applications, and their architectures are worth discussing more in depth.

## Large disks for any device

BrowserPod is intended to be a generic WebAssembly-powered in-browser sandbox. As such it goes beyond running one single program and must allow users to execute a diverse set of applications, depending on their workload. As a practical example, we include git and all its utilities in the default disk images and they will be useful to many users, but we don’t want _every_ user to download all this additional code unconditionally.

Real-world code can also access _thousands_ of small files in normal operation. Npm-based applications are one of the worst offenders in this regard. Storing files as plain HTTP resources it’s possible but at this scale it becomes not practical due to the overhead of each request.

> The BrowserPod filesystem implementation has been designed to carefully balance these aspects. These are the keys ideas:

Reviewer: ```suggestion
The [BrowserPod filesystem](https://browserpod.io/docs/understanding-browserpod/filesystem) implementation has been designed to carefully balance these aspects. These are the key ideas:
```

- **Filesystem:** POSIX-compatible, effectively an Ext2 implementation
- **Block-based streaming:** The disk image is not downloaded at startup, but progressively and on-demand.
- **Multiple backends:** Disk images can be stored on any HTTP server, with blocks being accessed using the standard HTTP byte ranges feature. For maximum performance a WebSocket service (hosted on Cloudflare), is used for our root images.

> - **Local-only storage:** Blocks downloaded will be cached locally using either the Origin Private File System (OPFS) API or IndexedDB. Any change to the blocks will only be saved locally, providing privacy-preserving persistence for BrowserPod sandboxes.

Reviewer: ```suggestion
- **Local-only storage:** Blocks downloaded will be cached locally using either the _Origin Private File System (OPFS) API_ or _IndexedDB_. Any changes to the blocks will be saved only locally, providing privacy-preserving persistence for BrowserPod sandboxes.
```

Ext2 could be considered a legacy choice, but it is simple to implement and works well at scale. We plan to add features over time to reach an Ext3 and eventually an Ext4 (i.e. modern) feature set.

## Solving networking, again

While the browser might intuitively seem a network-native platform, it’s not. There is no browser API to create raw TCP / UDP sockets in the browser, with the [Direct Socket API](https://developer.chrome.com/docs/iwa/direct-sockets) only available in _Isolated Web Apps_.

BrowserPod is in many ways a direct descendant of [WebVM](https://webvm.io), our x86 Debian virtual machine running in the browser. In the context of WebVM our solution was to [integrate with Tailscale](https://labs.leaningtech.com/blog/webvm-virtual-machine-with-networking-via-tailscale). This works, but it’s not a turnkey solution, and it requires the user to install the Tailscale client on their device.

For BrowserPod we wanted to go one step further and make sockets work out-of-the-box. Not only that, we also wanted to make it possible to access servers inside the sandbox from anywhere on the internet with zero setup. To explain, let’s start with the latter feature.

### Portals

> BrowserPod introduces _Portals_, automatically generated random domains that route HTTP traffic to services inside the sandbox. A portal URL will look like this

Reviewer: ```suggestion
BrowserPod introduces [_Portals_](https://browserpod.io/docs/understanding-browserpod/portals), automatically generated random domains that route HTTP traffic to services inside the sandbox. A portal URL will look like this
```

			<span class="text-bg-500">https://</span>
			<span class="text-sky-600 dark:text-sky-400">rbmhk1uf4vrqwns4mizrouzchdbljdohkssfp0mktzrijk4algta</span>
			<span class="text-amber-600 dark:text-amber-400">-4000</span>
			<span class="text-bg-500">.browserportal.io</span>
			<span class="text-emerald-600 dark:text-emerald-400">/path/index.html</span>
		</code>
	</pre>
	<figcaption class="text-center">
		Example of a typical portal URL: <span class="text-sky-600 dark:text-sky-400">blue</span> is the portal ID, <span class="text-amber-600 dark:text-amber-400">amber</span> is the internal port, and <span class="text-emerald-600 dark:text-emerald-400">green</span> is the request path.
	</figcaption>
</figure>

We adopted Cloudflare workers to support inbound traffic, so it was natural to use them as proxies for outbound traffic too, with a relatively straightforward implementation.

This solution works very well, but it introduces some thorny additional problems to prevent users from abusing our platform: it’s never a good idea to have an open proxy to the whole internet. To make BrowserPod “just work” while preventing abuse we settled on a tradeoff:

> - **Limited whitelisted domains for the free tier**: BrowserPod can be used by anybody with a few clicks and offers a generous free tier, but free users can only access a few selected and whitelisted domains. The list is expected to change over time but at its core it offers access to npm package registries and major git hosting platforms such as GitHub.

Reviewer: ```suggestion
- **Limited whitelisted domains for the free tier**: BrowserPod can be used by anybody with a few clicks and offers a generous free tier (see [pricing](https://browserpod.io/pricing/)), but free users can only access a few selected and whitelisted domains. The list is expected to change over time, but at its core, it offers access to npm package registries and major Git hosting platforms such as GitHub.
```

## Right, but what can I do with this?

The first version of BrowserPod was released in February 2026, but the platform is already very powerful. As of today you can run node, python, bash, git and many other Linux command line utilities in the sandbox.

We have prepared a few demos to showcase how the capabilities of BrowserPod can be applied to a few concrete scenarios, but the potential use cases are much, much broader than what we can show here.

### A learning platform for python

[SOMETHING BASIC WITH BLOCKING TERMINAL INPUT? POPEN?]

### Explore git repos from Web apps

[GIT CLONE + GIT SHOW -p WITH USER SELECTABLE REPO]

### Full-stack node applications in the browser

[[EXPRESS.JS](http://EXPRESS.JS) HELLO WORLD WITH PORTAL, LINK TO THE CONSOLE FOR MUCH MORE, LIST ALL THE FRAMEWORKS]

## Current limitations and future directions

As powerful as BrowserPod is, there are currently some significant limitations that are worth discussing. Good news though: we believe that we will be able to implement solutions for most of them in the medium term.

> - **Source code is required**: Our C/C++ compiler (Cheerp) can compile many applications and libraries that target Linux with little to no modification. We also expect the amount of required changes to shrink over time. This implies that in the a very broad set of packages could be compiled to run on BrowserPod, potentially reaching the same level of completeness as your average Linux distribution. The tradeoff here is that access to the source code is required. First-party proprietary applications can be compiled as well, but third-party binaries are currently unsupported. We plan to bridge this gap by integrating BrowserPod with our x86-to-WebAssembly virtualization engine, called [CheerpX](https://cheerpx.io). CheerpX is the engine that powers [WebVM](https://webvm.io), and it has been tested at scale over the last several years. While native WebAssembly binaries can be expected to run at almost native speed, some amount of slowdown is to be expected when virtualizing x86. This said we think the current performance of CheerpX will be sufficient for many practical use cases.

Reviewer: These bullet point texts seem too long to be bullet points. Would be more presentable to have each point's main idea as a subtitle and then the text organised as normal paragraphs.

## Building for the [TARGET] target

TODO

## Give it a try

To get an immediate feeling for what BrowserPod can do, check out [BrowserCode](https://browsercode.io/claude): Claude Code running unmodified in the browser on top of BrowserPod.

You can log-in with your Anthropic credentials and run the full agentic loop in your browser. All data stay local.

BrowserCode is [FOSS](https://github.com/leaningtech/browsercode) and it’s intended to be a reference integration of BrowserPod designed to showcase its capabilities. You can fork the repo to build your version, or something else entirely, by logging on the [BrowserPod console](https://github.com/leaningtech/browsercode).

You will be able to test BrowserPod on multiple ready-made sandboxes running many popular Node.js frameworks (Express.js, Next.js, Nuxt, React, Svelte, …), command line tools, and build your first integration using the BrowserPod API.

> All BrowserPod accounts have access to a generous free tier to experiment with the technology. If you plan to use BrowserPod for your Open Source project, or startup, please get in touch. We offer extended token grants especially for these cases.

Reviewer: ```suggestion
All BrowserPod accounts have access to a generous free tier to experiment with the technology. If you plan to use BrowserPod for your Open Source project or startup, [please get in touch](https://browserpod.io/pricing/). We offer extended token grants especially for these cases.
```
